File: src/python/eval_injection.py
"""CWE-94: Code injection via eval / exec."""

from flask import Flask, request

app = Flask(__name__)


@app.route("/calc")
def calc():
    expr = request.args.get("expr", "0")
    return str(eval(expr))


@app.route("/run", methods=["POST"])
def run():
    code = request.data.decode()
    ns = {}
    exec(code, ns)
    return repr(ns.get("result"))


if __name__ == "__main__":
    app.run(port=4007)

File: src/javascript/eval-injection.js
// CWE-94: Code injection via eval / Function constructor.

const express = require('express');
const app = express();

app.get('/calc', (req, res) => {
  const expr = req.query.expr;
  const result = eval(expr);
  res.send(String(result));
});

app.get('/run', (req, res) => {
  const code = req.query.code;
  const fn = new Function(code);
  res.send(String(fn()));
});

app.get('/template', (req, res) => {
  const tpl = req.query.t;
  const rendered = new Function('return `' + tpl + '`')();
  res.send(rendered);
});

app.listen(3007);
